### PR TITLE
Civic Bosch steering enhancements

### PR DESCRIPTION
**Pull request by jpancotti**
add steering enhancements for Civic Bosch:
-limit steer max to a00 to stay outside stock fw deadzone
-add support for modded fw

### CODE REVIEW

File: selfdrive/car/honda/values.py [hunk 1]
```diff
@@ -382,15 +382,14 @@ class CAR:
       b'39990-TBA-C020\x00\x00',
       b'39990-TBA-C120\x00\x00',
       b'39990-TGG-A020\x00\x00',
+      b'39990-TGG,A020\x00\x00',
       b'39990-TGG-A120\x00\x00',
```
Comment (csouers) [inline selfdrive/car/honda/values.py]: There is no A020, but A120 for the cars that currently have modded eps support. Line 385 should be `b'39990-TGG,A120\x00\x00',`

File: selfdrive/car/honda/interface.py [hunk 2]
```diff
@@ -181,9 +181,19 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), has_relay=False,
...
       ret.steerRatio = 15.38  # 10.93 is end-to-end spec
-      ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 4096], [0, 4096]] # TODO: determine if there is a dead zone at the top end
+      if eps_modified:
+        # stock request input values:     0x0000, 0x0067, 0x0107, 0x01CB, 0x0294, 0x035E, 0x0457, 0x060D, 0x06EE (STEER_CONFIG_INDEX: 1)
+        # stock request output values:    0x0000, 0x0380, 0x0800, 0x0C00, 0x0EB6, 0x10AE, 0x1200, 0x1200, 0x1200
+        # modified request output values: 0x0000, 0x0380, 0x0800, 0x0C00, 0x0EB6, 0x10AE, 0x1200, 0x1B00, 0x2D00
+        # stock filter output values:     0x00c0, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a
+        # modified filter output values:  0x00c0, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x0400, 0x0480
+        # note: max request allowed is 4096, but request is capped at 3840 in firmware, so modifications result in 2x max
+        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 2564, 8000], [0, 2564, 3840]]
```
Comment (gregjhogan) [inline selfdrive/car/honda/interface.py]: 2564 should be 2566?
```
>>> hex(((2566 * 0xddb4) >> 0xf) >> 2)
'0x457'
```
```suggestion
        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 2566, 8000], [0, 2566, 3840]]
```

File: selfdrive/car/honda/interface.py [hunk 3]
```diff
@@ -181,9 +181,19 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), has_relay=False,
...
+        ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.4], [0.12]]
+      else:
+        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 2564], [0, 2564]]
```
Comment (gregjhogan) [inline selfdrive/car/honda/interface.py]: 2564 should be 2566?
```
>>> hex(((2566 * 0xddb4) >> 0xf) >> 2)
'0x457'
```
```suggestion
        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 2566], [0, 2566]]
```

File: selfdrive/car/honda/interface.py [hunk 4]
```diff
@@ -181,9 +181,19 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), has_relay=False,
...
+        ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.8], [0.24]]
```
Comment (gregjhogan) [inline selfdrive/car/honda/interface.py]: seems like kp/ki should be raised by the amount that the request range was shortened (see nidec civic)
```suggestion
        ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[1.1], [0.33]]
```

File: selfdrive/car/honda/interface.py [hunk 5]
```diff
@@ -181,9 +181,19 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), has_relay=False,
...
-      ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 4096], [0, 4096]] # TODO: determine if there is a dead zone at the top end
+      if eps_modified:
+        # stock request input values:     0x0000, 0x0067, 0x0107, 0x01CB, 0x0294, 0x035E, 0x0457, 0x060D, 0x06EE (STEER_CONFIG_INDEX: 1)
+        # stock request output values:    0x0000, 0x0380, 0x0800, 0x0C00, 0x0EB6, 0x10AE, 0x1200, 0x1200, 0x1200
+        # modified request output values: 0x0000, 0x0380, 0x0800, 0x0C00, 0x0EB6, 0x10AE, 0x1200, 0x1B00, 0x2D00
+        # stock filter output values:     0x00c0, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a
+        # modified filter output values:  0x00c0, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x011a, 0x0400, 0x0480
+        # note: max request allowed is 4096, but request is capped at 3840 in firmware, so modifications result in 2x max
+        ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 2564, 8000], [0, 2564, 3840]]
+        ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.4], [0.12]]
```
Comment (gregjhogan) [inline selfdrive/car/honda/interface.py]: As the request range is increased, kp/ki normally need to be lowered.  kp is too high if you disturb the wheel and get it oscillating and it does not quickly stabilize at high speed.  Below suggestion is what we found best for nidec civic.
```suggestion
        ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.3], [0.1]]
```

File: selfdrive/car/honda/interface.py [hunk 6]
```diff
@@ -181,9 +181,19 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), has_relay=False,
       ret.wheelbase = CivicParams.WHEELBASE
       ret.centerToFront = CivicParams.CENTER_TO_FRONT
       ret.steerRatio = 15.38  # 10.93 is end-to-end spec
-      ret.lateralParams.torqueBP, ret.lateralParams.torqueV = [[0, 4096], [0, 4096]] # TODO: determine if there is a dead zone at the top end
+      if eps_modified:
+        # stock request input values:     0x0000, 0x0067, 0x0107, 0x01CB, 0x0294, 0x035E, 0x0457, 0x060D, 0x06EE (STEER_CONFIG_INDEX: 1)
+        # stock request output values:    0x0000, 0x0380, 0x0800, 0x0C00, 0x0EB6, 0x10AE, 0x1200, 0x1200, 0x1200
+        # modified request output values: 0x0000, 0x0380, 0x0800, 0x0C00, 0x0EB6, 0x10AE, 0x1200, 0x1B00, 0x2D00
```
Comment (gregjhogan) [inline selfdrive/car/honda/interface.py]: does stock firmware clips request at `0x67F`? if yes, `0x2D00` is more than 2x torque at `0x67F`
```
>>> import numpy as np
>>> np.interp([0x67f], [0x060D, 0x06EE], [0x1B00, 0x2D00])array([9246.72])
>>> hex(9246>>1)
'0x120f'
```

**Comment by csouers** [comment]
Will be doing testing on this VS what the original PI values were this week. Back to back, same time of day, same route, and with measure steering accuracy script logs. From my initial testing, using the nidec civic PI values on the hatchback are actually worse than what Joe had originally.

**Comment by jpancotti** [comment]
Accuracy testing for shortened request range on bosch civic thanks to discord user jad:
stock 0.7.4:
angle:  0 | error: 0.16 | =: 15% | +:  39% | -: 44% | count: 481996
angle:  1 | error: 0.17 | =: 14% | +:  26% | -: 59% | count: 370763
angle:  2 | error: 0.20 | =: 12% | +:  35% | -: 51% | count: 136276
angle:  3 | error: 0.23 | =: 11% | +:  39% | -: 48% | count: 57672
angle:  4 | error: 0.27 | =: 10% | +:  40% | -: 48% | count: 43897
angle:  5 | error: 0.37 | =:  7% | +:  37% | -: 54% | count: 20784
angle:  6 | error: 0.49 | =:  8% | +:  31% | -: 59% | count: 15127
angle:  7 | error: 1.01 | =:  3% | +:  20% | -: 76% | count: 4568
angle:  8 | error: 1.33 | =:  2% | +:   9% | -: 88% | count: 3734
angle:  9 | error: 1.98 | =:  2% | +:   2% | -: 94% | count: 1928
angle: 10 | error: 3.15 | =:  0% | +:   0% | -: 99% | count: 1862
angle: 11 | error: 4.06 | =:  0% | +:   0% | -: 99% | count: 1006
angle: 12 | error: 5.05 | =:  0% | +:   0% | -:100% | count: 1107
shortened request to a00, kp 1.1, ki 0.33:
angle:  0 | error: 0.19 | =: 12% | +:  39% | -: 47% | count: 467720
angle:  1 | error: 0.21 | =: 11% | +:  24% | -: 64% | count: 367627
angle:  2 | error: 0.25 | =:  9% | +:  34% | -: 56% | count: 127996
angle:  3 | error: 0.30 | =:  8% | +:  32% | -: 59% | count: 48508
angle:  4 | error: 0.34 | =:  8% | +:  32% | -: 58% | count: 36137
angle:  5 | error: 0.53 | =:  7% | +:  27% | -: 64% | count: 16355
angle:  6 | error: 0.84 | =:  5% | +:  14% | -: 80% | count: 11483
angle:  7 | error: 1.53 | =:  0% | +:   3% | -: 96% | count: 4251
angle:  8 | error: 2.55 | =:  0% | +:   1% | -: 98% | count: 3097
angle:  9 | error: 3.38 | =:  0% | +:   0% | -: 99% | count: 1722
angle: 10 | error: 4.38 | =:  0% | +:   0% | -:100% | count: 1701
angle: 11 | error: 5.22 | =:  0% | +:   0% | -:100% | count: 1091
angle: 12 | error: 5.86 | =:  0% | +:   0% | -:100% | count: 1081

steering error seems to be worse with the shortened range. added a third step to request range interp shorten ramp up from a00 to f00. needs testing.

File: selfdrive/car/honda/values.py (hunk 1 above)
Comment (bit-chemist) [reply selfdrive/car/honda/values.py:394]: I have a 2020 Civic Hatchback Sport Touring and my EPS fw version is `b'39990-TGG-A020\x00\x00'`

**Pull request closed without merging**